File: src/engine/adapters/mongo/definition.ts
import { EJSON } from 'bson';
import type { Db } from 'mongodb';
import type { DocumentSchemaMeta, ObjectDefinition } from '../../../shared/domain/definition';
import { encodePath, type NodePath } from '../../../shared/domain/tree';
import { collectionOptions } from './catalog';

const NO_OPTIONS_NOTE =
  'This collection has no creation options set (no validator, no capped-collection settings, etc).';

// Relaxed mode is deliberate (P19 D12): strict EJSON would render a schema's `minimum: 5` as
// `{"$numberInt":"5"}` in a pane whose whole job is readability, and a $jsonSchema is plain JSON
// by construction, so nothing is lost by relaxing it.
function stringifyRelaxed(value: unknown): string {
  return EJSON.stringify(value, undefined, 2, { relaxed: true });
}

// `$jsonSchema` renders as the Validation section's field table; any other validator document
// (e.g. `{ age: { $gte: 18 } }`) is real and genuinely allowed by Mongo — it renders as read-only
// JSON instead, never as an empty table pretending the schema fit a shape it doesn't have.
function buildDocumentSchema(
  validator: unknown,
  validationLevel: unknown,
  validationAction: unknown,
): DocumentSchemaMeta {
  const level = typeof validationLevel === 'string' ? validationLevel : null;
  const action = typeof validationAction === 'string' ? validationAction : null;
  if (typeof validator !== 'object' || validator === null) {
    return {
      validator: null,
      isJsonSchema: false,
      validationLevel: level,
      validationAction: action,
    };
  }
  const obj = validator as Record<string, unknown>;
  if ('$jsonSchema' in obj) {
    return {
      validator: stringifyRelaxed(obj.$jsonSchema),
      isJsonSchema: true,
      validationLevel: level,
      validationAction: action,
    };
  }
  return {
    validator: stringifyRelaxed(obj),
    isJsonSchema: false,
    validationLevel: level,
    validationAction: action,
  };
}

export async function buildDefinition(
  db: Db,
  segments: NodePath['segments'],
  databaseName: string,
  collection: string,
): Promise<ObjectDefinition> {
  const options = await collectionOptions(db, collection);
  const hasOptions = options !== undefined && Object.keys(options).length > 0;

  return {
    path: encodePath(segments),
    kind: 'collection',
    qualifiedName: `${databaseName}.${collection}`,
    language: 'json',
    statements: [hasOptions ? stringifyRelaxed(options) : '{}'],
    origin: 'server',
    notes: hasOptions ? [] : [NO_OPTIONS_NOTE],
    constraints: [],
    documentSchema: buildDocumentSchema(
      options?.validator,
      options?.validationLevel,
      options?.validationAction,
    ),
    sections: [],
    generatedAt: new Date().toISOString(),
  };
}

File: apps/kira-studio/internal/adapters/mongo/definition.go
package mongo

import (
	"bytes"
	"context"
	"encoding/json"
	"time"

	"go.mongodb.org/mongo-driver/v2/bson"
	mongodriver "go.mongodb.org/mongo-driver/v2/mongo"

	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/storage/model"
)

const noOptionsNote = "This collection has no creation options set (no validator, no capped-collection settings, etc)."

// stringifyRelaxed is definition.ts's stringifyRelaxed: relaxed mode is deliberate (P19 D12) —
// strict EJSON would render a schema's `minimum: 5` as `{"$numberInt":"5"}` in a pane whose whole
// job is readability, and a $jsonSchema is plain JSON by construction, so nothing is lost by
// relaxing it. Go's json.Indent does the 2-space pretty-printing MarshalExtJSON's own indent
// parameter would otherwise duplicate effort with.
func stringifyRelaxed(value any) (string, error) {
	compact, err := bson.MarshalExtJSON(value, false, false)
	if err != nil {
		return "", err
	}
	var buf bytes.Buffer
	if err := json.Indent(&buf, compact, "", "  "); err != nil {
		return "", err
	}
	return buf.String(), nil
}

func bsonDGet(d bson.D, key string) (any, bool) {
	for _, e := range d {
		if e.Key == key {
			return e.Value, true
		}
	}
	return nil, false
}

// buildDocumentSchema ports definition.ts's buildDocumentSchema: $jsonSchema renders as the
// Validation section's field table; any other validator document is real and genuinely allowed by
// Mongo — it renders as read-only JSON instead, never as an empty table pretending the schema fit
// a shape it doesn't have.
func buildDocumentSchema(validator any, validationLevel, validationAction any) (model.DocumentSchemaMeta, error) {
	var level, action *string
	if s, ok := validationLevel.(string); ok {
		level = &s
	}
	if s, ok := validationAction.(string); ok {
		action = &s
	}

	validatorDoc, ok := validator.(bson.D)
	if !ok {
		return model.DocumentSchemaMeta{IsJSONSchema: false, ValidationLevel: level, ValidationAction: action}, nil
	}
	if jsonSchema, ok := bsonDGet(validatorDoc, "$jsonSchema"); ok {
		text, err := stringifyRelaxed(jsonSchema)
		if err != nil {
			return model.DocumentSchemaMeta{}, err
		}
		return model.DocumentSchemaMeta{Validator: &text, IsJSONSchema: true, ValidationLevel: level, ValidationAction: action}, nil
	}
	text, err := stringifyRelaxed(validatorDoc)
	if err != nil {
		return model.DocumentSchemaMeta{}, err
	}
	return model.DocumentSchemaMeta{Validator: &text, IsJSONSchema: false, ValidationLevel: level, ValidationAction: action}, nil
}

// buildDefinition ports definition.ts's buildDefinition.
func buildDefinition(ctx context.Context, db *mongodriver.Database, segments []model.PathSegment, databaseName, collection string) (model.ObjectDefinition, error) {
	options, err := collectionOptions(ctx, db, collection)
	if err != nil {
		return model.ObjectDefinition{}, err
	}
	hasOptions := len(options) > 0

	statement := "{}"
	notes := []string{noOptionsNote}
	if hasOptions {
		text, err := stringifyRelaxed(options)
		if err != nil {
			return model.ObjectDefinition{}, err
		}
		statement = text
		notes = []string{}
	}

	var validator, validationLevel, validationAction any
	if hasOptions {
		validator, _ = bsonDGet(options, "validator")
		validationLevel, _ = bsonDGet(options, "validationLevel")
		validationAction, _ = bsonDGet(options, "validationAction")
	}
	documentSchema, err := buildDocumentSchema(validator, validationLevel, validationAction)
	if err != nil {
		return model.ObjectDefinition{}, err
	}

	return model.ObjectDefinition{
		Path:           model.EncodePath(segments),
		Kind:           "collection",
		QualifiedName:  databaseName + "." + collection,
		Language:       "json",
		Statements:     []string{statement},
		Origin:         "server",
		Notes:          notes,
		Constraints:    []model.ConstraintMeta{},
		DocumentSchema: &documentSchema,
		Sections:       []model.DefinitionSection{},
		GeneratedAt:    time.Now().UTC().Format(time.RFC3339Nano),
	}, nil
}
